src/Functions/DateTimeTransforms.h

```diff
@@ -111,19 +118,37 @@ struct ToStartOfDayImpl
     //TODO: right now it is hardcoded to produce DateTime only, needs fixing later. See date_and_time_type_details::ResultDataTypeMap for deduction of result type example.
     static inline UInt32 execute(const DecimalUtils::DecimalComponents<DateTime64> & t, const DateLUTImpl & time_zone)
     {
-        return time_zone.toDate(static_cast<time_t>(t.whole));
+        if (t.whole < 0)
+            return 0;
```

Comment:
What about something like -0.3? It's negative but 0 in the whole part.

Reply:
Good note, thanks